script/buildtim.sh

```diff
@@ -244,11 +244,17 @@ CLOCKSFILE=$CLOCKSPATH/clocks_ddr.txt
 # All DDR use the configuration for 800M
 DDRTYPE=`$GETDDRPARAMS $DDRTOPFILE ddr_type`
 
-$CLOCKSPATH/ddr_tool -i $DDRTOPFILE -o $CLOCKSPATH/ddr_static.txt
+DDRPRSET=`$GETDDRPARAMS $DDRTOPFILE ddr_preset_conf`
+if [ -z "$DDRPRSET" ]; then
+	$CLOCKSPATH/ddr_tool -i $DDRTOPFILE -o $CLOCKSPATH/ddr_static.txt
 
-if [ $? -ne 0 ]; then
-	echo "DDR_tool fails to run!"
-	exit 1
+	if [ $? -ne 0 ]; then
+		echo "DDR_tool fails to run!"
+		exit 1
+	fi
+else
+	# using preset ddr config file to instead of generated by ddr_tool
+	cp $CLOCKSPATH/$DDRPRSET $CLOCKSPATH/ddr_static.txt
```

Comment:
So this patch is replacing autogenerated `ddr_static.txt` file by pregenerated file `espressobin-ddr4-1cs-1g.txt` or `espressobin-ddr4-2cs-2g.txt`...

Reply:
... which looks like just a **hack**.

But rather it should be fixed generator of that file. It is in https://github.com/MarvellEmbeddedProcessors/mv-ddr-marvell/ repository, in subdirectory [a3720](https://github.com/MarvellEmbeddedProcessors/mv-ddr-marvell/blob/mv-ddr-devel/a3700/).

In file https://github.com/MarvellEmbeddedProcessors/mv-ddr-marvell/blob/mv-ddr-devel/a3700/mv_ddr_tim.h are basically all these commands which are put into autogenerated file `ddr_static.txt`. See sections `ddr4_1cs_tim` `ddr4_2cs_tim`.

During compilation of ARM trusted firmware is `ddr_static.txt` file autogenerated into directory `A3700-utils-marvell/tim/ddr/`

So I would suggest to look at differences between `espressobin-ddr4-1cs-1g.txt` (and then also `espressobin-ddr4-2cs-2g.txt`) from this pull request and `ddr_static.txt` file autogenerated in `A3700-utils-marvell/tim/ddr/` during ARM trusted firmwae compilation.

What are differences between these files?

And then create a pull request into mv-ddr-marvell repository and patch https://github.com/MarvellEmbeddedProcessors/mv-ddr-marvell/blob/mv-ddr-devel/a3700/mv_ddr_tim.h file to generate `ddr_static.txt` to match what is currently in `espressobin-ddr4-1cs-1g.txt` and `espressobin-ddr4-2cs-2g.txt`

Reply:
@heaterC: would you generate diff between those files?

Reply:
@pali: I had a look at the suggested files and I seems to me that `espressobin-ddr4-1cs-1g.txt` contains many more "commands" then the according section in [https://github.com/MarvellEmbeddedProcessors/mv-ddr-marvell/blob/mv-ddr-devel/a3700/mv_ddr_tim.h](https://github.com/MarvellEmbeddedProcessors/mv-ddr-marvell/blob/mv-ddr-devel/a3700/mv_ddr_tim.h). Besides I would not know how to split the "commands" into the two static chars `ddr4_1cs_tim_pre[]` and `ddr4_1cs_tim_post[]`.

Reply:
Can you show differences between `ddr_static.txt` from DDR4 1CS 1G variant and `espressobin-ddr4-1cs-1g.txt` files?

Reply:
This is the diff result between the contents of `mv_ddr_tim.h` (regarding ddr4-cs1) and `espressobin-ddr4-1cs-1g.txt`:
```
--- mv_ddr_tim.h.txt	2020-11-17 17:40:14.000000000 +0100
+++ espressobin-ddr4-1cs-1g.txt	2020-11-17 17:39:04.000000000 +0100
@@ -18,6 +18,25 @@
 WRITE: 0xC0000050 0x000001FF
 WRITE: 0xC000004C 0x00000000
 WRITE: 0xC0000054 0x00000480
+WRITE: 0xC0000300 0x00000B0C
+WRITE: 0xC0000380 0x00061A80
+WRITE: 0xC0000384 0x00027100
+WRITE: 0xC0000388 0x00000050
+WRITE: 0xC000038C 0x00000400
+WRITE: 0xC0000390 0x00800200
+WRITE: 0xC0000394 0x011803CF
+WRITE: 0xC0000398 0x01200255
+WRITE: 0xC000039C 0x00000808
+WRITE: 0xC00003A0 0x04050500
+WRITE: 0xC00003A4 0x00000002
+WRITE: 0xC00003A8 0x00001808
+WRITE: 0xC00003AC 0x1C250C1A
+WRITE: 0xC00003B0 0x0C0C060C
+WRITE: 0xC00003B4 0x05040602
+WRITE: 0xC00003B8 0x00000605
+WRITE: 0xC0000200 0x000E0001
+WRITE: 0xC0000204 0x00000000
+WRITE: 0xC0000220 0x05020635
 WRITE: 0xC00003C0 0x00020205
 WRITE: 0xC00003C4 0x00000003
 WRITE: 0xC00003DC 0x00081239
@@ -50,7 +69,8 @@
 DELAY: 0x0000000A
 
 ;Step9: DDRPHY Driver/Receiver & DQS internal Pullup/Pulldown settings
-WRITE: 0xC0001004 0xD0133449
+;WRITE: 0xC0001004 0xD0133449
+WRITE: 0xC0001004 0xD0677449
 WRITE: 0xC0001008 0xC770055A
 WRITE: 0xC000100C 0x5461DF77
 
@@ -78,6 +98,11 @@
 WRITE: 0xC00010D4 0x00300C03
 WRITE: 0xC00010D8 0x00300C03
 WRITE: 0xC00010DC 0x00300C03
+WRITE: 0xC00010E0 0x00010008
+WRITE: 0xC00010D0 0x000C0300
+WRITE: 0xC00010D4 0x00300C03
+WRITE: 0xC00010D8 0x00339CE7
+WRITE: 0xC00010DC 0x00300C03
 WRITE: 0xC00010E0 0x00010010
 WRITE: 0xC00010D0 0x000C0300
 WRITE: 0xC00010D4 0x00300C03
@@ -91,7 +116,7 @@
 STORE_SM_ADDR: SM0 0xC0001038
 LOAD_SM_ADDR: SM0 0xC0001038
 AND_SM_VAL: SM0 0xC0FFFFFF
-OR_SM_VAL: SM0 0x24000000
+OR_SM_VAL: SM0 0x32000000
 STORE_SM_ADDR: SM0 0xC0001038
 LOAD_SM_ADDR: SM0 0xC0001038
 AND_SM_VAL: SM0 0x3FFFFFFF
@@ -149,4 +174,4 @@
 
 AND_VAL: 0xC0000308 0xFFFFFFEF
 WRITE: 0xC0000028 0x11004000
-DELAY: 0x00000001
\ No newline at end of file
+DELAY: 0x00000001
```

Reply:
Please be aware that I have neglected wether a change belongs to variable ..._pre[] or ..._post[] .

Reply:
Perfect! There are only few changes. Can you locate where are those patterns in `ddr4_1cs_tim_pre[]` and `ddr4_1cs_tim_post[]` arrays?

Reply:
I believe that `static char ddr4_1cs_tim_post[]`starts originally with `"WRITE: 0xC00003C0 0x00020205\n"`. So maybe all lines above this belong to `ddr4_1cs_tim_pre[]`?

Reply:
You can test it. Edit that array, compile ATF again and it would generate new ddr_static.txt file. Whole purpose of those arrays are to generate ddr_static.txt. So the only thing which is needed is to figure out how to modify that two arrays to generate ddr_static.txt file to be same as espressobin-ddr4-1cs-1g.txt.

Reply:
@heaterC: Do you need some help with it? I see that you closed this pull request.

Reply:
I am trying to compile with modified `mv_ddr_tim.h` and when successfull, I will open a PR at `mv-ddr-marvell` repo.

Reply:
Great! When you would be opening a new pull request, mention me via "@pali" string in pull request description, so github send me notification about it.

Reply:
In branch `mv-ddr-devel`, something must be broken. Whenever I use this branch to compile, the u-boot image is unusable. `mv_ddr-armada-18.12` seems to work. I am using Armbian builder.

Reply:
The patch seems to work, I applied it also to `mv_ddr-armada-18.12`. I can't check `ddr_static.txt` directly though because I don't know how to only compile DDR4-cs1-800MHz, so when compilation of next configuration (DDR4-cs2 e.g.) starts, the file is already overwritten.

Reply:
Armbian is known to have outofdated binaries and custom patches which just cause problems (like erasal of MAC addresses, etc.)... Check that you have latest U-Boot 2020.10. Or use official instructions for building U-Boot and ARM trusted firmware:

https://gitlab.denx.de/u-boot/u-boot/-/blob/master/doc/README.marvell
https://trustedfirmware-a.readthedocs.io/en/latest/plat/marvell/armada/build.html

Reply:
I used latest U-Boot 2021.01-rc2 and no patches from Armbian. When choosing branch `mv-ddr-devel`, my EspressoBin seems to fall back to its backup u-boot image (2017.something, 17.06-wtmi). Took me a while to figure out what's happening and I am still not sure.

Reply:
To compile uboot:
```
$ make CROSS_COMPILE=aarch64-linux-gnu- mvebu_espressobin-88f3720_defconfig u-boot.bin
```
It will produce u-boot.bin

To compile ARM trusted firmware for DDR4 1CS 1GB 800MHz (depends on compiled u-boot binary and other repositories):
```
$ make PLAT=a3700 USE_COHERENT_MEM=0 DDR_TOPOLOGY=5 CLOCKSPRESET=CPU_800_DDR_800 CROSS_COMPILE=aarch64-linux-gnu- CROSS_CM3=arm-linux-gnueabi- BL33=/path/to/u-boot.bin WTP=/path/to/A3700-utils-marvell/ MV_DDR_PATH=/path/to/mv-ddr-marvell/ CRYPTOPP_PATH=/path/to/cryptopp/ mrvl_flash
```
This will produce final image `build/a3700/release/boot-image.bin` suitable for flashing and directory `build/a3700/release/uart-images/` with images suitable for booting via USB UART.

Reply:
And to boot via USB UART, there is wiki page with instructions:
http://wiki.espressobin.net/tiki-index.php?page=Bootloader+recovery+via+UART

Reply:
Does libcrypt work instead of crytopp?

Reply:
no, those are different projects

Reply:
@heaterC: I think I found bug in current mv-ddr-devel code. I think by default twin die should be set to not combined. Can you try following patch? It just change order of mv_ddr_twin_die enum, so `NOT_COMBINED` would be zero - default value.
```diff
diff --git a/ddr_topology_def.h b/ddr_topology_def.h
index 8a7e16a..af2021d 100644
--- a/ddr_topology_def.h
+++ b/ddr_topology_def.h
@@ -107,8 +107,8 @@ SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #define MV_DDR_MAX_IFACE_NUM	1
 
 enum mv_ddr_twin_die {
-	COMBINED,
 	NOT_COMBINED,
+	COMBINED,
 };
 
 struct bus_params {
```

Reply:
I compiled successfully with your patch and `mv-ddr-devel` branch! Boot output is:
```
TIM-1.0
WTMI-devel-18.12.1-0967979
WTMI: system early-init

DDR topology parameters:
========================
ddr type               DDR4
ddr speedbin           11
bus width              16-bits
cs num                 1
  cs[0] - group num    0
  cs[0] - bank num     8
  cs[0] - capacity     1024MiB
SVC REV: 5, CPU VDD voltage: 1.027V

DRAM windows:
=============
WIN[0] - base addr     0x60000000
WIN[0] - size          0x40000000

memory test region:
===================
CS[0]                  0x60000000 - 0x9fffffff

SELF-REFRESH TEST PASS
VREF READ TRAINING PASSED
VREF WRITE TRAINING PASSED
DLL TUNING PASSED
NOTICE:  Booting Trusted Firmware
NOTICE:  BL1: v1.5(debug):1f8ca7e (Marvell-devel-18.12.2)
NOTICE:  BL1: Built : 12:37:48, Nov 21 2020
NOTICE:  BL1: Booting BL2
NOTICE:  BL2: v1.5(debug):1f8ca7e (Marvell-devel-18.12.2)
NOTICE:  BL2: Built : 12:37:48, Nov 21 2020
NOTICE:  BL1: Booting BL31
NOTICE:  BL31: v1.5(debug):1f8ca7e (Marvell-devel-18.12.2)
NOTICE:  BL31: Built : 12:37:49

U-Boot 2021.01-rc2-armbian (Nov 21 2020 - 12:37:11 +0000)

DRAM:  1 GiB
Comphy-0: USB3_HOST0    5 Gbps    
Comphy-1: PEX0          2.5 Gbps  
Comphy-2: SATA0         5 Gbps    
SATA link 0 timeout.
AHCI 0001.0300 32 slots 1 ports 6 Gbps 0x1 impl SATA mode
flags: ncq led only pmp fbss pio slum part sxs 
PCIE-0: Link down
MMC:   sdhci@d0000: 0
Loading Environment from SPIFlash... SF: Detected mx25u3235f with page size 256 Bytes, erase size 4 KiB, total 4 MiB
OK
Model: Globalscale Marvell ESPRESSOBin Board
Net:   eth0: neta@30000 [PRIME]
Hit any key to stop autoboot:  0 
=> 
```
I don't have a board with 2GB DDR4 memory though, so I can't do any testing (and haven't yet done any patching regarding the 2GB to be honest).
Should I make a PR for the 1GB DDR4 1cs patch anyhow?

Reply:
If you have extracted needed changes from `espressobin-ddr4-1cs-1g.txt` and put them into `ddr4_1cs_tim_pre[]`+`ddr4_1cs_tim_post[]` and they are working fine, then yes, please create pull request. Also if my above patch is required, then put it into pull request too. Nice job!

Reply:
Also, can you prepare patches for 2GB DDR4? I think that if you do same process as with 1GB DDR4, it does not matter that only 1GB DDR4 variant was tested.